File: google/main/src/core/query/AndTerms.java
package main.src.core.query;

import java.util.regex.Pattern;
import java.util.*;

import main.src.core.engine.InvertedIndex;
import main.src.core.structures.Document;

public class AndTerms extends Terms {

    private static Pattern pattern = Pattern.compile("( *[^\\+\\-\\w]|^)(\\w+)");

    public AndTerms(String expression) {
        super();
        this.collect(expression, pattern, 2);
    }

    public HashSet<Document> getResults(InvertedIndex index) {
        HashSet<Document> results = new HashSet<>();
        if (!this.keys.isEmpty()) {
            results.addAll(index.getDocumentsOfToken(this.keys.get(0)));
            for (int idx = 1; idx < this.keys.size(); idx++)
                results.retainAll(index.getDocumentsOfToken(this.keys.get(idx)));
        }
        return results;
    };

}

File: GoogleSharp/GoogleSharp/src/core/query/AndTerms.cs
// Standard Library
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

// Internal
using GoogleSharp.Src.Core.Engine;
using GoogleSharp.Src.Core.Structures;

namespace GoogleSharp.Src.Core.Query {
    public class AndTerms : Terms {
        private static readonly Regex Pattern = new Regex (@"( *[^\+\-\w]|^)(\w+)", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        public AndTerms (string expression) : base () {
            Collect (expression, Pattern, 2);
        }

        public override HashSet<Document> GetResults (IInvertedIndex index) {
            var results = new HashSet<Document> ();
            if (Tokens.Any ()) {
                results.UnionWith (index.GetDocumentsOfToken (Tokens[0]));
                Tokens.ForEach (t => results.IntersectWith (index.GetDocumentsOfToken (t)));
            }
            return results;
        }
    }
}
